src/s3/util/fileViewerManager.ts

```diff
@@ -78,86 +90,146 @@ export class S3FileViewerManager {
                     return
                 }
                 await this.removeAndCloseTab(activeTab)
-                await this.openTab(fileNode)
+                await this.openInReadMode(fileNode)
                 this._onDidChange.fire(activeTab.s3Uri)
             }
         })
+    }
 
-        this.window.onDidChangeVisibleTextEditors(editors => {
-            const editorSet = new Set(editors)
-            for (const value of this.activeTabs.values()) {
-                //if visible text editors don't contain a given S3Tab anymore,
-                //set the S3Tab.editor to undefined
-                if (value.editor) {
-                    if (!editorSet.has(value.editor)) {
-                        value.editor = undefined
-                    }
-                }
-            }
+    public async focusAndCloseTab(
+        uri: vscode.Uri,
+        editor?: vscode.TextEditor,
+        workspace = vscode.workspace
+    ): Promise<void> {
+        const doc = editor ? editor.document : await workspace.openTextDocument(uri)
+        await this.window.showTextDocument(doc, {
+            preview: false,
+            viewColumn: editor?.viewColumn,
         })
+        await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
+    }
 
-        vscode.workspace.onDidCloseTextDocument(async closedTextDoc => {
-            if (this.activeTabs.has(closedTextDoc.uri.fsPath)) {
-                const closedTab = this.activeTabs.get(closedTextDoc.uri.fsPath)
-                closedTab!.editor = undefined
-            }
-        })
+    public async getCharset(fileNode: S3FileNode): Promise<string> {
+        let headResponse
+        try {
+            headResponse = await fileNode.s3.getHeadObject({
+                bucketName: fileNode.bucket.name,
+                key: fileNode.file.key,
+            })
+        } catch (e) {
+            getLogger().error('S3FileViewer: Error calling getHeadObject, error: ', e)
+        }
+
+        let type
+        let charset
+        if (headResponse) {
+            type = mime.contentType(headResponse.ContentType!)
```

Comment:
- this could be `const` by using a ternary. Also saves several lines of code.
- the `ContentType!` assumption can be avoided by checking `if (headResponse?.ContentType)`
- can `(type as string)`  be avoided by using toString()?

full code sample (not tested):

```typescript
if (headResponse?.ContentType) {
    const type = mime.contentType(headResponse.ContentType)
    const charset = mime.charset((type ?? '').toString())
    return charset
}
return ''
```